Add SparseSeries.to_coo method, a single test and one example.

cottrell opened the pull request:
xref #4343
closes #8048

This passes locally run nosetests but is failing on Travis. I think the latest Travis CI changes might have caused failures. I was able to get master to pass on Travis last week but now it is failing.

cottrell committed:
Add SparseSeries.to_coo and from_coo methods for interaction with scipy.sparse.

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,38 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+                                            (2, 1, 'b', 0),
+                                            (2, 1, 'b', 1)])
+   
+   ss = s.to_sparse() # SparseSeries
```
show ``s`` here (just put it on a line by itself)

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,38 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+   
+   ss = s.to_sparse() # SparseSeries
+   
+   A, il, jl = ss.to_coo(ilevels=[0, 1], jlevels=[2, 3], sort_labels=True)
```
I would name the levels in your example (to make it more clear). Are you testing named levels as well?

cottrell replied:
I hadn't implemented named-level specification yet. Is there an obvious choice for how the function should look? I am not too fond of the Index.ix-style methods. Perhaps modifying the signature to something like this would make sense?

    SparseSeries.to_coo(ilevels=[(0, ), (1, )], levels=None, sort_labels=False)

So ilevels would take integer indices (like iloc) and levels would take labels (like .loc). And specifying both would be an error.

jreback replied:
I don't think you actually need to do anything. ``.get_level_number(..)`` will turn names into numbers. You ust need to test this. I wouldn't change the signature.

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,38 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+   A, il, jl = ss.to_coo(ilevels=[0, 1], jlevels=[2, 3], sort_labels=True)
+
+   A
+   A.todense()
```
maybe put these in separate blocks so you can put some comments (or you can use hash comments, whatever looks better)

jreback commented on pandas/sandbox/scipy_sparse/sparse_coo_demo.py:
```diff
@@ -0,0 +1,23 @@
+                                         (1, 1, 'b', 1),
+                                         (2, 1, 'b', 0),
+                                         (2, 1, 'b', 1)])
+
```
this file is not necessary

jreback commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,77 @@
+
+def _squish(s):
+    """ Uniquify s while preserving order. Elements of s must be appendable to set. """
+    seen = set()
```
You can do this
```
In [1]: l = [1,2,2,1,5,7]

In [2]: pd.Index(l).unique()
Out[2]: array([1, 2, 5, 7])
```

jreback commented on pandas/sparse/series.py:
```diff
@@ -654,6 +656,25 @@ def combine_first(self, other):
+
+        Parameters
+        ----------
+        ilevels: tuple/list
```
indicate that these are number/names of levels

jreback commented on pandas/sparse/tests/test_sparse.py:
```diff
@@ -742,6 +743,60 @@ def test_combine_first(self):
         assert_sp_series_equal(result, result2)
         assert_sp_series_equal(result, expected)
 
+    def test_to_coo(self):
```
create a new class to test with. then you can do the setup in ``setUp`` (just inherit from ``TestCase``) and will work

jreback commented on pandas/sparse/tests/test_sparse.py:
```diff
@@ -742,6 +743,60 @@ def test_combine_first(self):
+    def test_to_coo(self):
+
+        # test data generators for scipy_sparse interaction, move outside (the class) if the data ends up being reused
+        def _get_sparse_series_0():
```
You need to have a test for each of the exceptions that you raise

jreback commented on vb_suite/sparse.py:
```diff
@@ -37,3 +37,20 @@
+
+
+setup = common_setup + """
+s = pd.Series([3.0, nan, 1.0, 2.0, nan, nan])
```
maybe this much bigger (x 1000). otherwise not much of a tests

cottrell commented:
I have incorporated your comments and have reorganized the tests. I have also added a from_coo method. I can take the from_coo out if that is just complicating things. The vbench stuff is still untested unless it runs in Travis CI somehow.

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,83 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+   il
+   jl
+
+``pandas.sparse.series.from_coo``
```
If you do this: ``:meth:`~pandas.SparseSeries.from_coo` `` then these will show up as links to the API docs

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,83 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+
+.. ipython:: python
+
+   ss = pandas.sparse.series.from_coo(A)
```
hmm, this should be a class-method on ``SparseSeries.from_coo`` no?

jreback commented on pandas/sparse/series.py:
```diff
@@ -654,6 +656,46 @@ def combine_first(self, other):
+        A, il, jl = sparse_series_to_coo(self, ilevels, jlevels, sort_labels=sort_labels)
+        return(A, il, jl)
+
+def from_coo(A, dense_index=False):
```
so add ``@classmethod`` and put this in the ``SparseSeries``

jreback commented:
@cottrell looking pretty good. pls squash when you have a chance as well.

cottrell commented:
I had some weird, unreproducible issues with the squashed version not passing Travis CI but it seems to be ok now. For some reason, the SparseSeries.from_coo is not appearing in the API docs while the SparseSeries.to_coo is.

jorisvandenbossche commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,100 @@
+    whole = set(whole)
+    parts = [set(x) for x in parts]
+    if set.intersection(*parts) != set():
+        raise(ValueError('Is not a partition because intersection is not null.'))
```
No brackets around the ValueError are needed here (and some other occurences)

jorisvandenbossche commented:
Small comment: can you check for PEP8? (there are eg some lines that fall of the github screen because of the line length)

jorisvandenbossche commented on doc/source/api.rst:
```diff
@@ -621,6 +621,7 @@ Serialization / IO / Conversion
    Series.to_msgpack
    Series.to_json
    Series.to_sparse
+   SparseSeries.to_coo
```
Maybe add a section for methods specific to sparse? (just an idea, on the other hand it belongs to the `to_..` methods)

jorisvandenbossche commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,83 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+
+.. ipython:: python
+
+   s = pandas.Series([3.0, nan, 1.0, 3.0, nan, nan])
```
you can just use `pd` here instead of pandas (it is already imported)

jorisvandenbossche commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,83 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+   :suppress:
+
+   from numpy import nan
+   import scipy.sparse
```
I think the recommended way from scipy it to do `from scipy import sparse`, and then `sparse.coo_matrix` (although the docs seem to use `from scipy.sparse import coo_matrix`

jorisvandenbossche replied:
+ I would also show the import (don't suppress it), as it is a non standard import

jorisvandenbossche commented on pandas/sparse/series.py:
```diff
@@ -654,6 +656,47 @@ def combine_first(self, other):
+
+        Parameters
+        ----------
+        A: scipy.sparse.coo_matrix
```
Can you add an explanation of `dense_index` here as well?

jorisvandenbossche replied:
+ for sphinx/numpydoc, there should be a space after the kwarg (so a space before and after the colon) to render correctly (it is rather strict in these things). So it looks like:

```
kwarg : type
    Explanation.
```

jorisvandenbossche commented on pandas/sparse/series.py:
```diff
@@ -654,6 +656,47 @@ def combine_first(self, other):
+        s: SparseSeries
+        """
+        return(coo_to_sparse_series(A, dense_index=dense_index))
 # overwrite series methods with unaccelerated versions
```
can you add a blank line between above two lines?

cottrell replied:
Where did you mean here with the blank line? I reorganized and corrected this doc string anyway. I'll push shortly.

jorisvandenbossche replied:
I meant between line 699 and 700 above, so there is a blank line between the end of the function definition and the following thing (in fact, following PEP8, there could even be two blank lines, as it is the end of the class definition here, but that is a bit nitpicking)

jorisvandenbossche commented:
@cottrell I added a bunch of mostly documentation comments

cottrell commented:
I have added a more friendly message when to_coo is called with nlevels<2 (and a test for this Exception) and have incorporated comments from @jorisvandenbossche . In adding a section in the API, I wondered whether it might make sense to someday hook the to_coo method into the Series instance methods simply to aid discovery. It might simply be adding clutter.

jorisvandenbossche commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,108 @@
+TODO:
+    * more tests
+    * from_coo
+    * vbench
```
Are these todo's still valid?

jorisvandenbossche commented on pandas/sparse/series.py:
```diff
@@ -29,12 +29,14 @@
 
 from pandas.util.decorators import Appender
 
+from pandas.sparse.scipy_sparse import sparse_series_to_coo, coo_to_sparse_series
```
I think we can't just import this here, as scipy is only an optional dependency. 
Maybe putting it inside the method is enough? Or does it need a try/except?

cottrell replied:
I've moved "import scipy" inside the method. If you run without scipy you get an import error when you call the to_coo function ... I think this is probably ok behaviour.

jorisvandenbossche commented on pandas/sparse/tests/test_sparse.py:
```diff
@@ -30,6 +32,7 @@
 from pandas.sparse.api import (SparseSeries, SparseTimeSeries,
                                SparseDataFrame, SparsePanel,
                                SparseArray)
+import scipy.sparse
```
Also here the same comment. There should be some skipping of the tests when there is no scipy. Search for `_skip_if_no_scipy()` to see how it is done elsewhere

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,84 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+.. ipython:: python
+   :suppress:
+
+   from numpy import nan
```
kill the pd import and just use ``Series`` for example. BUT, do add the ``from numpy import nan`` line in the code (as you are directly using it, which is fine)

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +135,84 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+   ss
+
+Specifying ``dense_index=True`` will result in an index that is the Cartesian product of the
+row and columns coordinates of the matrix. Note that this will consume a significant amount of memory
```
hmm, conceptually I think we should *always* do ``dense_index=False`` followed by an explicity ``reindex``. What is the advantage of providing an option which people will use 'incorrectly' ?

cottrell replied:
The problem is that if dense_index=False always, you should really just use a Series (not a SparseSeries). If I am thinking about this correctly, coo_matrices simply store the shape and the non-null labels (coords) while SparseSeries carry the extra burden of storing ALL the labels (since they could be anything).

cottrell commented:
Incorporated the comments. I have noticed that pandas.sparse.scipy_sparse is appearing in the public modules (in tab completion IPython). It wasn't obvious to me what the best way of hiding this was.

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +134,81 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+
+The method requires a ``MultiIndex`` with two or more levels.
+
+.. ipython:: python
```
extra section here. pls kill

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +134,81 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+                                            names=['A', 'B', 'C', 'D'])
+
+   s
+   ss = s.to_sparse() # SparseSeries
```
pls put the comments above the statement (though this is just preference)

jreback commented on doc/source/sparse.rst:
```diff
@@ -135,3 +134,81 @@ recommend using ``block`` as it's more memory efficient. The ``integer`` format
+   A
+   A.todense()
+   il
+   jl
```
maybe show a slightly different version of the same conversion in addition, e.g. maybe not sorting or something (or doing more/less levels), just to give the causal reader 2 variations.

jreback commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+    labels respectively. Returns the sparse_matrix, row and column labels. """
+    if ss.index.nlevels < 2:
+        raise ValueError('to_coo requires MultiIndex with nlevels > 2')
+    if any(ss.index.to_series().duplicated()):
```
you can just call ``not ss.index.is_unique``

jreback commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+    s = s.sort_index()
+    s = s.to_sparse()  # TODO: specify kind?
+    if dense_index:
+        # is there a better constructor method to use here?
```
you can use ``MultiIndex.from_product([i,j])``

jreback commented on pandas/sparse/tests/test_sparse.py:
```diff
@@ -743,6 +744,124 @@ def test_combine_first(self):
 
 
+class TestSparseSeriesScipyInteraction(tm.TestCase):
+
```
pls add the issue number as a comment

jreback commented:
@cottrell pls add a release note. Make a note under Enhancements. Provide a short example (and provide a link to the new doc section).

Overall looks pretty good, some minor changes.

I think we might need to revisit the conversion algo (at a later time) if perf is an issue, can you post some perf numbers for a couple of cases (I know you have the vbench in there so that is good).

cottrell commented:
Updated and rebased (there was a conflict in the whatsnew section). Some quick perf numbers here:

http://nbviewer.ipython.org/github/cottrell/notebooks/blob/master/Performance%20scipy.sparse%20examples.ipynb

I forgot to note but I was running on a typical 2.3 Ghz i5 macbook.

Performance seems ok for moderate < 1e7 length indices but definitely could be done better. Could probably be improved along with the MultiIndex groupby operations as these are fairly similar activities.

jreback commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -106,6 +106,51 @@ Enhancements
 - ``tseries.frequencies.to_offset()`` now accepts ``Timedelta`` as input (:issue:`9064`)
 
 - ``Timedelta`` will now accept nanoseconds keyword in constructor (:issue:`9273`)
+- Added :meth:`SparseSeries.to_coo` and :meth:`SparseSeries.from_coo` methods
```
let's add this in its own sub-section (e.g. put ``-----``) underneath the title, then you don't have to indent at all

cottrell replied:
Done. Used ^^^^^^ for subparagraph or whatever that is.

jreback commented:
@cottrell minor change in docs

@jorisvandenbossche @shoyer 

lgtm, any other comments?

jreback commented:
Pls add a short example to the doc-string for ``to/from_coo`` 

Further, I think it might be instructive to have a small doc section that basically shows what this operation is from a reshaping point of view. e.g. using your whatsnew example (not that though this is simple it first densifies so only should do with a relatively small matrix otherwise your memory will blow up; your comprehensive impl is better).

```
In [54]: ss.to_dense().unstack(level=['C','D'])
Out[54]: 
C     a       b    
D     0   1   0   1
A B                
1 1 NaN NaN   1   3
  2   3 NaN NaN NaN
2 1 NaN NaN NaN NaN

In [55]: ss.to_dense().unstack(level=['C','D']).fillna(0).values
Out[55]: 
array([[ 0.,  0.,  1.,  3.],
       [ 3.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.]])

In [56]: ss.to_dense().unstack(level=['C','D']).fillna(0).index.values
Out[56]: array([(1, 1), (1, 2), (2, 1)], dtype=object)

In [57]: ss.to_dense().unstack(level=['C','D']).fillna(0).columns.values
Out[57]: array([('a', 0), ('a', 1), ('b', 0), ('b', 1)], dtype=object)

In [58]: A.todense()
Out[58]: 
matrix([[ 3.,  0.,  0.,  0.],
        [ 0.,  0.,  1.,  3.],
        [ 0.,  0.,  0.,  0.]])
```

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+
+    v, i, j, il, jl = _to_ijv(
+        ss, ilevels=ilevels, jlevels=jlevels, sort_labels=sort_labels)
+    import scipy.sparse
```
Better to import scipy.sparse at the top of the function so things fail faster if the import isn't available -- it's also slightly more standard, though I can't find it in PEP8

cottrell replied:
I moved this into the function as per @jorisvandenbossche comment above (not sure how to reference the comment). Should I move it back? Idea was that scipy is an optional dependency.

shoyer replied:
I just meant move it right below the docstring for this function (line 70 above)

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+    # to keep things simple, only rely on integer indexing (not labels)
+    ilevels = [ss.index._get_level_number(x) for x in ilevels]
+    jlevels = [ss.index._get_level_number(x) for x in jlevels]
+    ss = ss.copy()
```
do you really want to copy the entire sparse series here? That could be expensive.

shoyer commented on pandas/sparse/series.py:
```diff
@@ -654,6 +656,48 @@ def combine_first(self, other):
         dense_combined = self.to_dense().combine_first(other)
         return dense_combined.to_sparse(fill_value=self.fill_value)
 
+    def to_coo(self, ilevels=(0,), jlevels=(1,), sort_labels=False):
```
What about using `row_levels` and `column_levels` here instead of `ilevels`/`jlevels`? I like more explicit names.

shoyer commented on pandas/sparse/series.py:
```diff
@@ -654,6 +656,48 @@ def combine_first(self, other):
+        Returns
+        -------
+        y : scipy.sparse.coo_matrix
+        il : list (row labels)
```
Likewise, I would consider returns `rows`/`columns` instead of `il`/`jl`

shoyer replied:
Also, wouldn't it be more useful for this function to return a `MultiIndex` for the `rows`/`columns` rather than native Python lists? You can do lots of stuff to lookup values in a MultiIndex (e.g., `get_loc`/`get_indexer`/`slice_indexer`) even if you don't have a dataframe to store them in -- not so much with plain lists.

jreback replied:
I think return a list here is ok
it's a list of tuples (and if you have a multi index just mi.values); a list is more expects here outside of pandas

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+    if sort_labels:
+        labels = sorted(list(labels))
+    d = OrderedDict((k, i) for i, k in enumerate(labels))
+    return(d)
```
`return` is a control structure, not a function, so it's usually written `return d` not `return(d)`

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+
+def _get_label_to_i_dict(labels, sort_labels=False):
+    """ Return OrderedDict of unique labels to number. Optionally sort by label. """
+    labels = Index(map(tuple, labels)).unique().tolist()  # squish
```
I would consider just returning an MultiIndex from this function instead of an OrderedDict. The multi-index supports the dictionary queries you need with `get_loc`/`get_indexer` and will be much faster. Also, you might be able to simply/speedup the logic considerably by using pandas methods:
```
labels = Index(labels).drop_duplciates()
if sort_labels:
    labels = labels.order()
return labels
```

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+        values_ilabels = [tuple(x[i] for i in levels) for x in nonnull_labels]
+        labels_to_i = _get_index_subset_to_coord_dict(
+            ss.index, levels, sort_labels=sort_labels)
+        i_coord = [labels_to_i[i] for i in values_ilabels]
```
If `labels_to_i` is a MultiIndex instead of an OrderedDict (like I recommend above), then this is as simple as:
`i_coord = labels_to_i.get_indexer(values_ilabels)` -- and again, we're using the fast pandas methods, not a loop in pure Python. The only cavaet of using `get_indexer` is that missing values are marked by `-1` instead of raising `KeyError`, so you might need to check for that explicitly (if you haven't already guaranteed that can't happen).

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+
+
+def _get_index_subset_to_coord_dict(index, subset, sort_labels=False):
+    ilabels = list(zip(*[index.get_level_values(i) for i in subset]))
```
to do this in a vectorized way, I think this might suffice:
```
new_index = pd.MultiIndex([index.levels[i] for i in subset],
                          [index.labels[i] for i in subset])
```

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+
+    def get_indexers(levels):
+        """ Return sparse coords and dense labels for subset levels """
+        values_ilabels = [tuple(x[i] for i in levels) for x in nonnull_labels]
```
I think you could also vectorize this line using the MultiIndex constructor, as above, if you made `nonnull_labels` a MultiIndex.

shoyer commented:
I have a strong suspicion you can optimize many of these internal steps by using a MultiIndex. That would speed things up a lot (and simply much of the logic).

Otherwise, the API looks reasonable, though I would use variables with meaningful names like `rows`/`columns` instead of `i`/`j` and return MultiIndexes.

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,102 @@
+    blocs = ss._data.values.sp_index.blocs
+    blength = ss._data.values.sp_index.blengths
+    nonnull_labels = list(
+        itertools.chain(*[ss.index.values[i:(i + j)] for i, j in zip(blocs, blength)]))
```
If this is the slow step (which seems likely), you could write a little routine to do this in a loop in Cython.

shoyer commented:
Obviously, performance is not strictly necessary here for this first draft but it's something to think about.

shoyer commented:
Just to clarify, you don't need to vectorize everything here according to my suggestions. For a first draft, slow is better that not at all.

cottrell commented:
I will hopefully soon vectorize as I think it will make it more readable. I sort of feel this is whole thing in its cleanest form should really be just a trivial application of a groupby on the levels but I am hitting a problem that I think I hit before which would push me down another path. Does anyone know if the following is a bug? I searched quickly but did not find anything. The problem seems to be only for groupby on index (with levels):

    In [13]: a
    Out[13]:
    a  b  c  d
    1  2  a  0    0
             1    1
       1  b  0    2
             1    3
    2  1  b  0    4
             1    5
    dtype: int64

    In [14]: a.groupby(level=['a', 'b'], sort=True).first()
    Out[14]:
    a  b
    1  1    2
       2    0
    2  1    4
    dtype: int64

    In [15]: a.groupby(level=['a', 'b'], sort=False).first() # should not have sorted keys
    Out[15]:
    a  b
    1  1    2
       2    0
    2  1    4
    dtype: int64

shoyer commented:
@cottrell can you should how you created `a` in this example? Also, `a.index.levels` and `a.index.labels` would likely demystify this .

cottrell commented:
@shoyer sorry forgot to include the first bit. Have updated.

cottrell commented:
I have incorporated some of the comments from @shoyer which has shortened code quite a bit. This included the patch at https://github.com/pydata/pandas/issues/9444. There are some TODO to review ... worried a bit about relying on the behaviour of some pandas operations for certain steps now (for example dropna on sparse series).

Note: I still need to add examples to the docstrings.

cottrell commented:
Performance was too bad using the groupby method so I have reverted to the older python method and left the groupby method commented out (lines 56-57 of scipy_sparse.py).

Some timing results here suggest it is some overhead in the groupby methods that is the difference. 
http://nbviewer.ipython.org/github/cottrell/notebooks/blob/master/pandas%20scipy.sparse%20performance.ipynb

Please let me know if you have any ideas about MultiIndex groupby performance. I have not tried yet, but a diff of pstats output might be useful.

shoyer commented on pandas/sparse/scipy_sparse.py:
```diff
@@ -0,0 +1,138 @@
+    # from the SparseSeries: get the labels and data for non-null entries
+    values = ss._data.values._valid_sp_values
+
+    # TODO: is replacing this code with dropna below safe?
```
Yes, I think this is safe.

cottrell replied:
Thanks, I've dropped that chunk of code and the TODO.

jreback commented:
merged via https://github.com/pydata/pandas/commit/9ac01a7ab02727f69c70401cc4bb8432fe70b83f

@cottrell thank you very much for this!

pls review built docs (give about an hour for travis to build). Pls submit a follow up pr if needed (or for any changes that you think are necessary; or open an issue).

jorisvandenbossche commented:
@cottrell There is an error in the docs:

```
>>>-------------------------------------------------------------------------

Exception in /tmp/doc/source/sparse.rst at block ending on line 195

Specify :okexcept: as an option in the ipython:: block to suppress this message

---------------------------------------------------------------------------

ValueError Traceback (most recent call last)

<ipython-input-36-a2b3f3f12d26> in <module>()

1 A, rows, columns = ss.to_coo(row_levels=['A', 'B', 'C'],

2 column_levels=['D'],

----> 3 sort_labels=False)

/home/travis/build/pydata/pandas/pandas/sparse/series.pyc in to_coo(self, row_levels, column_levels, sort_labels)

709 """

710 A, rows, columns = _sparse_series_to_coo(

--> 711 self, row_levels, column_levels, sort_labels=sort_labels)

712 return A, rows, columns

713

/home/travis/build/pydata/pandas/pandas/sparse/scipy_sparse.pyc in _sparse_series_to_coo(ss, row_levels, column_levels, sort_labels)

113

114 v, i, j, rows, columns = _to_ijv(

--> 115 ss, row_levels=row_levels, column_levels=column_levels, sort_labels=sort_labels)

116 sparse_matrix = scipy.sparse.coo_matrix(

117 (v, (i, j)), shape=(len(rows), len(columns)))

/home/travis/build/pydata/pandas/pandas/sparse/scipy_sparse.pyc in _to_ijv(ss, row_levels, column_levels, sort_labels)

90

91 i_coord, i_labels = get_indexers(row_levels)

---> 92 j_coord, j_labels = get_indexers(column_levels)

93

94 return values, i_coord, j_coord, i_labels, j_labels

/home/travis/build/pydata/pandas/pandas/sparse/scipy_sparse.pyc in get_indexers(levels)

80

81 labels_to_i = _get_index_subset_to_coord_dict(

---> 82 ss.index, levels, sort_labels=sort_labels)

83 #######################################################################

84 #######################################################################

/home/travis/build/pydata/pandas/pandas/sparse/scipy_sparse.pyc in _get_index_subset_to_coord_dict(index, subset, sort_labels)

74 ilabels, sort_labels=sort_labels)

75 labels_to_i = Series(labels_to_i)

---> 76 labels_to_i.index = MultiIndex.from_tuples(labels_to_i.index)

77 labels_to_i.index.names = [index.names[i] for i in subset]

78 labels_to_i.name = 'value'

/home/travis/build/pydata/pandas/pandas/core/index.pyc in from_tuples(cls, tuples, sortorder, names)

3604 tuples = tuples.values

3605

-> 3606 arrays = list(lib.tuples_to_object_array(tuples).T)

3607 elif isinstance(tuples, list):

3608 arrays = list(lib.to_object_array_tuples(tuples).T)

/home/travis/build/pydata/pandas/pandas/lib.so in pandas.lib.tuples_to_object_array (pandas/lib.c:54298)()

ValueError: Buffer dtype mismatch, expected 'Python object' but got 'long'

<<<-------------------------------------------------------------------------
```

cottrell commented:
@jorisvandenbossche

Yes, this is addressed here:

https://github.com/pydata/pandas/pull/9583

I was building docs only with python 3 so didn't catch this on my setup.
